src/hotspot/share/opto/type.cpp

```diff
@@ -1572,215 +1575,120 @@ const TypeInteger* TypeInteger::minus_1(BasicType bt) {
-const TypeInt *TypeInt::UBYTE;  // Unsigned Bytes, 0 to 255
-const TypeInt *TypeInt::CHAR;   // Java chars, 0-65535
-const TypeInt *TypeInt::SHORT;  // Java shorts, -32768-32767
-const TypeInt *TypeInt::POS;    // Positive 32-bit integers or zero
-const TypeInt *TypeInt::POS1;   // Positive 32-bit integers
-const TypeInt *TypeInt::INT;    // 32-bit integers
-const TypeInt *TypeInt::SYMINT; // symmetric range [-max_jint..max_jint]
-const TypeInt *TypeInt::TYPE_DOMAIN; // alias for TypeInt::INT
-
-//------------------------------TypeInt----------------------------------------
-TypeInt::TypeInt( jint lo, jint hi, int w ) : TypeInteger(Int, w), _lo(lo), _hi(hi) {
-}
+const TypeInt* TypeInt::MAX;    // INT_MAX
+const TypeInt* TypeInt::MIN;    // INT_MIN
+const TypeInt* TypeInt::MINUS_1;// -1
+const TypeInt* TypeInt::ZERO;   // 0
+const TypeInt* TypeInt::ONE;    // 1
+const TypeInt* TypeInt::BOOL;   // 0 or 1, FALSE or TRUE.
+const TypeInt* TypeInt::CC;     // -1,0 or 1, condition codes
+const TypeInt* TypeInt::CC_LT;  // [-1]  == MINUS_1
+const TypeInt* TypeInt::CC_GT;  // [1]   == ONE
+const TypeInt* TypeInt::CC_EQ;  // [0]   == ZERO
+const TypeInt* TypeInt::CC_NE;
+const TypeInt* TypeInt::CC_LE;  // [-1,0]
+const TypeInt* TypeInt::CC_GE;  // [0,1] == BOOL (!)
+const TypeInt* TypeInt::BYTE;   // Bytes, -128 to 127
+const TypeInt* TypeInt::UBYTE;  // Unsigned Bytes, 0 to 255
+const TypeInt* TypeInt::CHAR;   // Java chars, 0-65535
+const TypeInt* TypeInt::SHORT;  // Java shorts, -32768-32767
+const TypeInt* TypeInt::NON_ZERO;
+const TypeInt* TypeInt::POS;    // Positive 32-bit integers or zero
+const TypeInt* TypeInt::POS1;   // Positive 32-bit integers
+const TypeInt* TypeInt::INT;    // 32-bit integers
+const TypeInt* TypeInt::SYMINT; // symmetric range [-max_jint..max_jint]
+const TypeInt* TypeInt::TYPE_DOMAIN; // alias for TypeInt::INT
 
-//------------------------------make-------------------------------------------
-const TypeInt *TypeInt::make( jint lo ) {
-  return (TypeInt*)(new TypeInt(lo,lo,WidenMin))->hashcons();
+TypeInt::TypeInt(const TypeIntPrototype<jint, juint>& t, int w, bool dual)
+  : TypeInteger(Int, t.normalize_widen(w), dual), _lo(t._srange._lo), _hi(t._srange._hi),
+    _ulo(t._urange._lo), _uhi(t._urange._hi), _bits(t._bits) {
+  DEBUG_ONLY(t.verify_constraints());
 }
 
-static int normalize_int_widen( jint lo, jint hi, int w ) {
-  // Certain normalizations keep us sane when comparing types.
-  // The 'SMALLINT' covers constants and also CC and its relatives.
-  if (lo <= hi) {
-    if (((juint)hi - lo) <= SMALLINT)  w = Type::WidenMin;
-    if (((juint)hi - lo) >= max_juint) w = Type::WidenMax; // TypeInt::INT
-  } else {
-    if (((juint)lo - hi) <= SMALLINT)  w = Type::WidenMin;
-    if (((juint)lo - hi) >= max_juint) w = Type::WidenMin; // dual TypeInt::INT
+const Type* TypeInt::try_make(const TypeIntPrototype<jint, juint>& t, int w, bool dual) {
+  auto new_t = t.canonicalize_constraints();
+  if (!new_t._present) {
```

Comment:
Here `empty` would read a little more naturally, just an idea.
Ah, and I would rename `new_t` -> `canonical_t` or `canonicalized_t`.